### PR TITLE
GuaranteedUpdate must write if stored data is not canonical

### LINKED ISSUE #48393: Storage version migration is broken due to an optimization in GuaranteedUpdate
Pull https://github.com/kubernetes/kubernetes/pull/35415 optimized guaranteed update to avoid a Get against etcd3.  Instead of reading the data that etcd has, it reserializes the in memory version of the object.  The problem is that we use the delta between the etcd version and the new version we are writing to get rid of out of date versions (i.e. going from v1beta3 to v1, we switched the apiserver config to v1, then issued a no-op write to all objects).  With the optimization in place, the only way to force a write is to make a concrete change to the serialized object, like adding an annotation.

Until we fix this, we can't drop an old API version, and someone updating to protobuf can't for instance ensure all their data is protobuf at rest.

### PR DESCRIPTION
**Pull request by smarterclayton**
An optimization added to the GuaranteedUpdate loop changed the
comparison of the current objects serialization against the stored data,
instead comparing to the in memory object, which defeated the mechanism
we use to migrate stored data (GET then PUT should update the version stored in etcd if the canonical serialization has changed)

This commit preserves that optimization but correctly verifies the in
memory serialization against the on disk serialization by fetching the
latest serialized data. Since most updates are not no-ops, this should
not regress the performance of the normal path.

Fixes #48393

```release-note
When performing a GET then PUT, the kube-apiserver must write the canonical representation of the object to etcd if the current value does not match. That allows external agents to migrate content in etcd from one API version to another, across different storage types, or across varying encryption levels. This fixes a bug introduced in 1.5 where we unintentionally stopped writing the newest data.
```

### CODE REVIEW

**Comment by wojtek-t** [comment]
LGTM - thanks @smarterclayton !

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 1]
```diff
@@ -297,6 +299,21 @@ func (s *store) GuaranteedUpdate(
...
+			if mustCheckData {
+				getResp, err := s.client.KV.Get(ctx, key, s.getOps...)
+				if err != nil {
+					return err
+				}
+				origState, err = s.getState(getResp, key, v, ignoreNotFound)
+				if err != nil {
+					return err
+				}
+				mustCheckData = false
```
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: if `bytes.Equal(origState.data, neworigState.data) == true` here, can we return directly?

https://github.com/kubernetes/kubernetes/pull/54780

**Pull request merged**